### PR TITLE
[Data] Improve filesystem retry coverage

### LINKED ISSUE #43803: [Data] Retry on `OSError: AWS Error NETWORK_CONNECTION during GetObject operation: curlCode: 28, Timeout was reached`
### What happened + What you expected to happen

ray.data.read_parquet_bulk crashes on parquet_base_datasource's pq.read_table with
```
OSError: AWS Error NETWORK_CONNECTION during GetObject operation: curlCode: 28, Timeout was reached
```
This error should be retried

### Versions / Dependencies

ray==2.9.2

### Reproduction script

```
large_dataset = ... # (e.g., s3 directory with 100k parquet files of 1GiB each of arbitrary data)
data = ray.data.read_parquet_bulk(large_dataset)
data.write_datasink(...) # write to s3
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

See https://github.com/ray-project/ray/issues/43803#issuecomment-1985691360.

## Related issue number

Fixes https://github.com/ray-project/ray/issues/43803

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/context.py [hunk 1]
```diff
@@ -93,6 +94,15 @@
...
 )
 
+DEFAULT_RETRIED_FILESYSTEM_ERRORS = (
+    "AWS Error INTERNAL_FAILURE",
+    "AWS Error NETWORK_CONNECTION",
+    "AWS Error SLOW_DOWN",
+    "AWS Error UNKNOWN (HTTP status 503)",
+    "AWS Error ACCESS_DENIED",
+    "AWS Error NETWORK_CONNECTION",
+)
```
Comment (bveeramani) [inline python/ray/data/context.py:104]: Consolidating all of the transient errors in one place so users don't need to configure retries for opening files/reading files/writing files separately.

File: python/ray/data/context.py [hunk 2]
```diff
@@ -231,6 +241,9 @@ class DataContext:
             call is made with a S3 URI.
         wait_for_min_actors_s: The default time to wait for minimum requested
             actors to start before raising a timeout, in seconds.
+        retried_filesystem_errors: A list of substrings of error messages that should
+            trigger a retry when reading or writing files. This is useful for handling
+            transient errors when reading from remote storage systems.
```
Comment (raulchen) [inline python/ray/data/context.py:246]: Nit, `filesystem_errors` sounds is a bit too specific to the implementation details. 
A more general name (e.g. `io_errors`) would be better.

**Review by raulchen — APPROVED** [review]

File: python/ray/data/context.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/context.py:246]: Addressed

**Comment by anyscalesam** [comment]
@raulchen @bveeramani shouldn't we also add coverage for GCP specific ERRORs?

**Pull request merged**